File: src/services/neoid.rs
use serde::Deserialize;
use serde_json;

pub struct NeoIdClient {
    pub base_url: String,
    pub client_id: String,
    pub client_secret: String,
    client: reqwest::Client,
}

#[derive(Debug, Deserialize)]
pub struct NeoIdUser {
    pub id: String,
    pub email: String,
    pub display_name: Option<String>,
    pub avatar: Option<String>,
    pub first_name: Option<String>,
    pub last_name: Option<String>,
    pub username: Option<String>,
    pub role: String,
}

impl NeoIdUser {
    pub fn display_name_resolved(&self) -> String {
        if let Some(name) = &self.display_name {
            if !name.trim().is_empty() {
                return name.clone();
            }
        }
        if let Some(name) = &self.username {
            if !name.trim().is_empty() {
                return name.clone();
            }
        }
        let first = self.first_name.as_deref().unwrap_or("");
        let last = self.last_name.as_deref().unwrap_or("");
        let full = format!("{} {}", first, last).trim().to_string();
        if !full.is_empty() {
            return full;
        }
        self.email.split('@').next().unwrap_or("").to_string()
    }
}

#[derive(Deserialize)]
#[allow(dead_code)]
struct OAuthTokenData {
    access_token: Option<String>,
    refresh_token: Option<String>,
    id_token: Option<String>,
    token_type: Option<String>,
    expires_in: Option<u64>,
}

#[derive(Deserialize)]
#[allow(dead_code)]
struct OAuthTokenResponse {
    ok: bool,
    data: OAuthTokenData,
}

impl NeoIdClient {
    pub fn new(base_url: &str, client_id: &str, client_secret: &str) -> Self {
        Self {
            base_url: base_url.trim_end_matches('/').to_string(),
            client_id: client_id.to_string(),
            client_secret: client_secret.to_string(),
            client: reqwest::Client::builder()
                .timeout(std::time::Duration::from_secs(10))
                .build()
                .unwrap(),
        }
    }

    /// Build the OAuth2 authorize URL for the consent screen.
    pub fn build_authorize_url(
        &self,
        redirect_uri: &str,
        state: &str,
        code_challenge: Option<&str>,
        code_challenge_method: Option<&str>,
    ) -> String {
        let mut url = format!(
            "{}/api/v1/oauth/authorize?response_type=code&client_id={}&redirect_uri={}&state={}",
            self.base_url,
            urlencoding::encode(&self.client_id),
            urlencoding::encode(redirect_uri),
            urlencoding::encode(state),
        );
        if let Some(cc) = code_challenge {
            url.push_str(&format!("&code_challenge={}", urlencoding::encode(cc)));
        }
        if let Some(ccm) = code_challenge_method {
            url.push_str(&format!("&code_challenge_method={}", ccm));
        }
        url
    }

    /// Exchange authorization code for tokens.
    /// Uses the OAuth2 Token endpoint with `grant_type=authorization_code`.
    pub async fn exchange_auth_code(
        &self,
        code: &str,
        redirect_uri: &str,
        _code_verifier: Option<&str>,
    ) -> Result<OAuthTokenResult, String> {
        let url = format!("{}/api/v1/oauth/token", self.base_url);

        let body = serde_json::json!({
            "grant_type": "authorization_code",
            "code": code,
            "redirect_uri": redirect_uri,
            "client_id": self.client_id,
            "client_secret": self.client_secret,
        });

        eprintln!("[neoid] POST {} body: {}", url, serde_json::to_string_pretty(&body).unwrap_or_default());

        let resp = self
            .client
            .post(&url)
            .header("Content-Type", "application/json")
            .json(&body)
            .send()
            .await
            .map_err(|e| format!("neo id oauth token request failed: {}", e))?;

        let status = resp.status();
        let body_text = resp.text().await.unwrap_or_default();
        if !status.is_success() {
            eprintln!("[neoid] token exchange failed ({}): {}", status, body_text);
            return Err(format!("neo id oauth token error {}: {}", status, body_text));
        }
        eprintln!("[neoid] token exchange success");

        let wrapper: OAuthTokenResponse = serde_json::from_str(&body_text)
            .map_err(|e| format!("failed to parse neo id oauth token response: {}", e))?;

        let access_token = wrapper.data.access_token.unwrap_or_default();
        if access_token.is_empty() {
            return Err("neo id oauth token response missing access_token".to_string());
        }

        Ok(OAuthTokenResult {
            access_token,
            refresh_token: wrapper.data.refresh_token,
            id_token: wrapper.data.id_token,
            expires_in: wrapper.data.expires_in,
        })
    }

    /// Fetch user profile from NeoID using the access token.
    pub async fn get_profile(&self, access_token: &str) -> Result<NeoIdUser, String> {
        let url = format!("{}/api/v1/user/profile", self.base_url);
        let resp = self
            .client
            .get(&url)
            .header("Authorization", format!("Bearer {}", access_token))
            .send()
            .await
            .map_err(|e| format!("neo id profile request failed: {}", e))?;

        if !resp.status().is_success() {
            return Err("neo id profile request failed".to_string());
        }

        #[derive(Deserialize)]
        struct ProfileResponse {
            data: NeoIdUser,
        }

        let profile: ProfileResponse = resp
            .json()
            .await
            .map_err(|e| format!("failed to parse neo id profile response: {}", e))?;

        Ok(profile.data)
    }

    /// Notify NeoID that a user deleted their account.
    pub async fn notify_user_deleted(&self, user_id: &str) {
        if self.base_url.is_empty() || self.client_secret.is_empty() {
            return;
        }
        let url = format!("{}/api/v1/webhooks/user-deleted", self.base_url);
        let body = serde_json::json!({
            "event": "user.deleted",
            "user_id": user_id,
            "client_id": self.client_id,
        });
        let _ = self
            .client
            .post(&url)
            .header("Content-Type", "application/json")
            .json(&body)
            .send()
            .await;
    }
}

pub struct OAuthTokenResult {
    pub access_token: String,
    pub refresh_token: Option<String>,
    pub id_token: Option<String>,
    pub expires_in: Option<u64>,
}

File: src/services/neoid.ts
import { config } from "../config"

interface OAuth2TokenResponse {
  access_token: string
  token_type: string
  expires_in: number
  refresh_token: string
  id_token: string
}

interface NeoIdUserResponse {
  id: string
  email: string
  displayName?: string
  avatar?: string
  role: string
}

/** Neo ID wraps payloads as `{ ok: true, data: T }`. */
function unwrapData<T>(json: unknown): T {
  if (json && typeof json === "object" && "data" in json && (json as { ok?: boolean }).ok !== false) {
    return (json as { data: T }).data
  }
  return json as T
}

class NeoIdClient {
  private clientId = config.neoId.clientId
  private clientSecret = config.neoId.clientSecret

  /**
   * Builds the Neo ID authorize URL. Extra params (PKCE code_challenge,
   * state, …) are appended verbatim so SPAs can start a public-client flow
   * and exchange the code from the browser.
   */
  getAuthorizeUrl(redirectUri?: string, extra?: Record<string, string>): string {
    const redirect = redirectUri || config.neoId.redirectUri
    const params = new URLSearchParams({
      client_id: this.clientId,
      redirect_uri: redirect,
      response_type: "code",
      scope: config.neoId.scope,
    })
    for (const [key, value] of Object.entries(extra ?? {})) {
      if (value) params.set(key, value)
    }
    return `${config.neoId.authorizeUrl}?${params.toString()}`
  }

  async exchangeCode(code: string, redirectUri?: string): Promise<OAuth2TokenResponse> {
    const redirect = redirectUri || config.neoId.redirectUri
    const res = await fetch(config.neoId.tokenUrl, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        grant_type: "authorization_code",
        code,
        redirect_uri: redirect,
        client_id: this.clientId,
        client_secret: this.clientSecret,
      }),
    })

    if (!res.ok) {
      const text = await res.text()
      throw new Error(`Neo ID token exchange failed: ${res.status} ${text}`)
    }

    const json = unwrapData<Record<string, unknown>>(await res.json())
    return {
      access_token: (json.access_token || json.accessToken) as string,
      token_type: (json.token_type || "Bearer") as string,
      expires_in: Number(json.expires_in ?? 3600),
      refresh_token: (json.refresh_token || json.refreshToken) as string,
      id_token: (json.id_token || json.idToken) as string,
    }
  }

  async refreshTokens(refreshToken: string): Promise<OAuth2TokenResponse> {
    const res = await fetch(config.neoId.tokenUrl, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        grant_type: "refresh_token",
        refresh_token: refreshToken,
        client_id: this.clientId,
        client_secret: this.clientSecret,
      }),
    })

    if (!res.ok) {
      const text = await res.text()
      throw new Error(`Neo ID refresh failed: ${res.status} ${text}`)
    }

    const json = unwrapData<Record<string, unknown>>(await res.json())
    return {
      access_token: (json.access_token || json.accessToken) as string,
      token_type: (json.token_type || "Bearer") as string,
      expires_in: Number(json.expires_in ?? 3600),
      refresh_token: (json.refresh_token || json.refreshToken) as string,
      id_token: (json.id_token || json.idToken) as string,
    }
  }

  async getUser(accessToken: string): Promise<NeoIdUserResponse> {
    const res = await fetch(config.neoId.userInfoUrl, {
      headers: { Authorization: `Bearer ${accessToken}` },
    })

    if (!res.ok) {
      throw new Error(`Neo ID user fetch failed: ${res.status}`)
    }

    const profile = unwrapData<Record<string, unknown>>(await res.json())
    return {
      id: String(profile.id ?? profile.sub ?? ""),
      email: String(profile.email ?? ""),
      displayName: (profile.displayName ?? profile.name) as string | undefined,
      avatar: (profile.avatar ?? profile.picture) as string | undefined,
      role: String(profile.role ?? "user"),
    }
  }

  async listRefreshTokens(authHeader: string): Promise<{ id: string; deviceName?: string; createdAt: string }[]> {
    // Neo ID exposes active sessions, not a separate oauth2 token list
    const res = await fetch(`${config.neoId.issuer}/api/v1/sessions`, {
      headers: { Authorization: authHeader },
    })
    if (!res.ok) return []
    const sessions = unwrapData<{ id: string; deviceInfo?: string; createdAt: string }[]>(await res.json())
    return (sessions ?? []).map((s) => ({
      id: s.id,
      deviceName: s.deviceInfo,
      createdAt: s.createdAt,
    }))
  }

  async revokeRefreshToken(sessionId: string, authHeader: string): Promise<void> {
    await fetch(`${config.neoId.issuer}/api/v1/sessions/${sessionId}`, {
      method: "DELETE",
      headers: { Authorization: authHeader, "Content-Type": "application/json" },
    })
  }

  async revokeAllRefreshTokens(authHeader: string): Promise<void> {
    await fetch(`${config.neoId.issuer}/api/v1/sessions`, {
      method: "DELETE",
      headers: { Authorization: authHeader, "Content-Type": "application/json" },
    })
  }
}

export const neoid = new NeoIdClient()
